test/e2e/cluster_deletion.go

```diff
@@ -0,0 +1,399 @@
+	// if only one infrastructure provider exists, it will be used, otherwise the operation will fail if more than one exists.
+	InfrastructureProvider *string
+
+	// Flavor, if specified is the template flavor used to create the cluster for testing.
+	// If not specified, the default flavor for the selected infrastructure provider is used.
+	Flavor *string
+
+	// ControlPlaneMachineCount defines the number of control plane machines to be added to the workload cluster.
+	// If not specified, 1 will be used.
+	ControlPlaneMachineCount *int64
+
+	// WorkerMachineCount defines number of worker machines to be added to the workload cluster.
+	// If not specified, 1 will be used.
+	WorkerMachineCount *int64
+
+	// Allows to inject functions to be run while waiting for the control plane to be initialized,
+	// which unblocks CNI installation, and for the control plane machines to be ready (after CNI installation).
+	ControlPlaneWaiters clusterctl.ControlPlaneWaiters
+
+	// Allows to inject a function to be run after test namespace is created.
+	// If not specified, this is a no-op.
+	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
+
+	// Allows to inject a function to be run after machines are provisioned.
+	// If not specified, this is a no-op.
+	PostMachinesProvisioned func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace, workloadClusterName string)
+
+	// ClusterDeletionPhases define kinds which are expected to get deleted in phases.
+	ClusterDeletionPhases []ClusterDeletionPhase
+}
+
+type ClusterDeletionPhase struct {
+	// DeleteKinds are the DeleteKinds which are supposed to be deleted in this phase.
+	// Note: If Machines is added here, the Machine must have an owner with a kind from DeleteBlockKinds or DeleteKinds.
+	DeleteKinds sets.Set[string]
+	// DeleteBlockKinds are the kinds which are considered to be deleted in this phase, but need to be blocking.
+	// Note: If Machines is added here, the Machine must have an owner with a kind from DeleteBlockKinds or DeleteKinds.
+	DeleteBlockKinds sets.Set[string]
+	// objects will be filled later by objects which match a kind given in one of the above sets.
+	objects []client.Object
+}
+
+// ClusterDeletionSpec implements a test that verifies that MachineDeployment rolling updates are successful.
+// ClusterDeletionSpec goes through the following steps:
+// * Create a cluster.
+// * Add finalizer to objects defined in input.ClusterDeletionPhases.DeleteBlockKinds.
+// * Trigger deletion for the cluster.
+// * For each phase:
+//   - Verify objects expected to get deleted in previous phases are gone.
+//   - Verify objects expected to be in deletion have a deletionTimestamp set but still exist.
+//   - Verify objects expected to be deleted in a later phase don't have a deletionTimestamp set.
+//   - Remove finalizers for objects in deletion.
```

Comment:
These steps correlate to what is configured via DeleteKinds & DeleteBlockKinds, right?. Would it be helpful to make clear here which step uses which config field?